fuse_constraints/src/normal_delta_pose_2d.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, Clearpath Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_constraints/normal_delta_pose_2d.h>
+#include <fuse_core/util.h>
+
+#include <Eigen/Core>
+#include <glog/logging.h>
+
+
+namespace fuse_constraints
+{
+
+NormalDeltaPose2D::NormalDeltaPose2D(const fuse_core::MatrixXd& A, const fuse_core::Vector3d& b) :
+  A_(A),
+  b_(b)
+{
+  CHECK_GT(A_.rows(), 0);
+  CHECK_EQ(A_.cols(), 3);
+  set_num_residuals(A_.rows());
+}
+
+bool NormalDeltaPose2D::Evaluate(
+  double const* const* parameters,
+  double* residuals,
+  double** jacobians) const
+{
+  const auto R1_transpose = fuse_core::rotationMatrix2D(parameters[1][0]).transpose();  // orientation1
+  const auto position_delta =
```

Comment:
I did a bit of digging and found the `auto` keyword is causing me grief. You might have a better handle on Eigen's tendency to return expression templates instead of fully evaluated matrix objects. But, if I specify the type on these two lines instead of using `auto`, the unit tests pass. At least on my local machine.

```suggestion
  const fuse_core::Matrix2d R1_transpose = fuse_core::rotationMatrix2D(parameters[1][0]).transpose();  // orientation1
  const fuse_core::Vector2d position_delta =
```

Reply:
That lazy Eigen evaluation! I should forget about `auto` when it comes to Eigen. After all, the expectation is that it'd be evaluated when used below, but that means it'd probably get evaluated more than once!

I believe only the `position_delta` is lazy evaluated in this case. But I've followed your suggestion and changed both.